src/java/main/org/apache/zookeeper/server/NIOServerCnxnFactory.java

```diff
@@ -607,9 +618,28 @@ public static ByteBuffer getDirectBuffer() {
     // sessionMap is used by closeSession()
     private final ConcurrentHashMap<Long, NIOServerCnxn> sessionMap =
         new ConcurrentHashMap<Long, NIOServerCnxn>();
-    // ipMap is used to limit connections per IP
-    private final ConcurrentHashMap<InetAddress, Set<NIOServerCnxn>> ipMap =
-        new ConcurrentHashMap<InetAddress, Set<NIOServerCnxn>>( );
+    // ipMap is used to limit connections and connection rate per IP
+    private final ConcurrentHashMap<InetAddress, IpCnxns> ipMap
+        = new ConcurrentHashMap<InetAddress, IpCnxns>();
+
+    // for each IP, we keep a RateLimiter to limit connection rate,
+    // as well as a set of connections to limit total number of connections
+    private static class IpCnxns {
```

Comment:
Eventually, we need to implement same logic for NettyServerCnxnFactory so better to up this class to ServerCnxnFactory or make it a top level class. The former is fine, imho.